src/java.base/share/classes/java/io/FileDescriptor.java

```diff
@@ -59,15 +61,35 @@ public final class FileDescriptor {
     private List<Closeable> otherParents;
     private boolean closed;
 
+    /**
+     * Called by FileDispatcherImpl when the file descriptor is about to be closed natively.
+     */
+    public void markClosedByNIO() {
+        resource.closedByNIO = true;
+    }
+
     class Resource implements jdk.internal.crac.JDKResource {
+        private static final boolean COLLECT_FD_STACKTRACES =
+                GetBooleanAction.privilegedGetProperty(JDKContext.COLLECT_FD_STACKTRACES_PROPERTY);
+
+        private boolean closedByNIO;
```

Comment:
I think it may perfectly reside in FileDesriptor, as it looks like a bug for me that FileDispatcherImpl does not change valid() status of the FileDesriptor. Or even make FileDispatcherImpl to reset FileDescriptor.fd? Are there reasons not to do that?

Reply:
I haven't changed that to minimize the impact of changes; I still assume at one point it made sense to close the native FD without going through FileDescriptor. The reply I got on nio-dev was this:

> In the java.io APIs, FIS/FOS/RAF define getFD methods so the FileDescriptor can be obtained by applications. This is problematic for a bunch of reasons but for the discussion here, setting it to -1 is needed for the FileDescriptor::valid. Another point is that java.io pre-dates the more robust async close mechanism that is NIO and setting it to -1 helps avoid trying to use a file descriptor that has been closed and/or recycled.
> In the NIO area, FileDescriptor objects do not leak to applications. The only case where it is necessary to set to -1 is on Windows with sockets and this is because there is no equivalent of dup2 there. Another point here is that FileDescriptors aren't really needed as the channels implementation just need the raw file descriptor or handle. 

Having that in FileDescriptor seemed more consolidated but if you prefer it that way I won't object. So, should I do that or risk breaking stuff by setting fd to -1?

Reply:
From the reply [1] I understand that there was no reason to keep track of the native fd/valid() status in FileDescriptors used by NIO, so that was not done. And for me the status of the FD is closer to FD itself, but I don't have a strong preference.

[1] https://mail.openjdk.org/pipermail/nio-dev/2023-January/013075.html, just for bookkeeping